cpp/src/arrow/type.h

```diff
@@ -2151,4 +2151,32 @@ const std::vector<std::shared_ptr<DataType>>& IntervalTypes();
 ARROW_EXPORT
 const std::vector<std::shared_ptr<DataType>>& PrimitiveTypes();
 
+ARROW_EXPORT
+bool IsSignedIntType(std::shared_ptr<DataType>);
```

Comment:
I'm leaning towards -1 on adding these APIs. There are similar APIs, with different names and inlined, that take a type id. It's not significantly to write `IsSignedIntType(type)` rather than `is_signed_integer(type->id())`.

Reply:
In addition, the signature is inefficient as this is always copying a shared_ptr (implying an atomic reference increment). Instead this could just take a `const DataType&`.